<a href="https://colab.research.google.com/github/sreent/data-management-intro/blob/main/MySQL%20Hand-On%20Lab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **MySQL Hand-On Lab**

##Setting Up MySQL Environment

In [1]:
# install and start mysql server
!apt -qq update > /dev/null
!apt -y -qq install mysql-server > /dev/null
!/etc/init.d/mysql restart > /dev/null

# mount gdrive, required for downloaded dataset (or sql data dump)
import os
from google.colab import drive
drive.mount('/content/drive')

# in gdrive,  create a folder for storing sql data dump (a git repo in our case) of the employees sample database
%mkdir -p /content/drive/MyDrive/mysql

# clone a repo storing the employees data (the source data for sql data dump)
%cd /content/drive/MyDrive/mysql
# if the repo is already existed then pull the update. if not, then we clone the reop.
!git -C hr-schema-mysql pull || git clone https://github.com/nomemory/hr-schema-mysql.git hr-schema-mysql > /dev/null

# navigate to the downloaded repo
%cd /content/drive/MyDrive/mysql/hr-schema-mysql
# restore the employees sample database from the dump data
!mysql -t < hr-schema-mysql.sql > /dev/null

# back to the default directory
%cd /content

# create a database user and grant access to the employees database
!mysql -e "CREATE USER IF NOT EXISTS 'francis'@'localhost' IDENTIFIED WITH mysql_native_password BY 'california';" 
!mysql -e "GRANT ALL PRIVILEGES ON hr.* TO 'francis'@'localhost';"

# install sql magic, required for running sql commands in jupyter notebook
!pip install -q ipython-sql
# install mysql python connector
!pip install -q pymysql mysql-connector-python mysqlclient
# install sqlalchemy
!pip install -q sqlalchemy==1.4.4

# enable sql magic in the notebook
%reload_ext sql

# queries are returned in the pandas format 
%config SqlMagic.autopandas=True

import pandas as pd
# maximum # of rows to be rendered
pd.set_option('display.max_rows', 25)

# connect to the hr database
%sql mysql+pymysql://francis:california@localhost/hr





su: warning: cannot change directory to /nonexistent: No such file or directory
Mounted at /content/drive
/content/drive/MyDrive/mysql
Already up to date.
/content/drive/MyDrive/mysql/hr-schema-mysql
/content
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 25.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.7/43.7 kB 2.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 27.4/27.4 MB 39.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 88.1/88.1 kB 10.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 26.9 MB/s eta 0:00:00


##The Employees Schema
The Human Resources (HR) schema is part of the Oracle Sample  Schemas. The  practice labs in this course use data from the HR schema. Oracle provides several schema objects for training purposes.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1lbMXyFIaNQTjZPmDB3VghajrAAIm3FH0)

##**Lab 1**: Retrieving Data Using the SQL <code>SELECT</code> Statement
This practice covers the following topics:
- Selecting all data from different tables
- Describing the structure of tables
- Performing arithmetic calculations and specifying column names

**Task 1.1** The following <code>SELECT</code> statement executes successfully: (True/False)
<pre>
SELECT last_name, job_id, salary AS sal
FROM employees;
</pre>

In [2]:
%%sql
# Insert Your Code Here
SELECT last_name, job_id, salary AS sal
FROM employees;

 * mysql+pymysql://francis:***@localhost/hr
107 rows affected.


last_name      job_id       sal
0        King     AD_PRES  24000.00
1     Kochhar       AD_VP  17000.00
2     De Haan       AD_VP  17000.00
3      Hunold     IT_PROG   9000.00
4       Ernst     IT_PROG   6000.00
..        ...         ...       ...
102       Fay      MK_REP   6000.00
103    Mavris      HR_REP   6500.00
104      Baer      PR_REP  10000.00
105   Higgins      AC_MGR  12000.00
106     Gietz  AC_ACCOUNT   8300.00

[107 rows x 3 columns]

**Task 1.2** The following <code>SELECT</code> statement executes successfully: (True/False)
<pre>
SELECT *
FROM jobs;
</pre>

In [ ]:
%%sql
# Insert Your Code Here

**Task 1.3** There are 4 coding errors in the following statement. Can you identify and fix them?
<pre>
SELECT employee_id, last_name
       sal x 12 ANNUAL SALARY
FROM employees;
</pre>

In [ ]:
%%sql
# Insert Your Code Here

**Task 1.4** You have been hired as a SQL programmer. Your first task is to determine the structure of the <code>departments</code> table and its contents.

To determine the <code>departments</code> table structure:
<pre>
DESCRIBE departments
</pre>

![task-1.4a.png](https://drive.google.com/uc?id=13wod4XZZ5IB25emHUa8OeyLEb26uzqCC)

In [ ]:
%%sql
# Insert Your Code Here

To view the data contained in the <code>departments</code> table:
<pre>
SELECT *
FROM departments;
</pre>

![task-1.4b.png](https://drive.google.com/uc?id=19XzjmVZcJrZWpF4toYnkL53YjKs6UjZ0)


In [ ]:
%%sql
# Insert Your Code Here

**Task 1.5** Your task is to determine the structure of the <code>employees</code> table and its contents.

To determine the <code>employees</code> table structure:
<pre>
DESCRIBE employees
</pre>

![task-1.5.png](https://drive.google.com/uc?id=12Oi1uc_XZJChKNOU4uNon3W53dVFt6CI)

In [ ]:
%%sql
# Insert Your Code Here

**Task 1.6** The HR department wants a query to display the last name, hire date, and
employee ID for each employee, with the employee ID appearing first. Provide an alias <code>start_date</code> for the <code>hire_date</code> column.
<pre>
SELECT employee_id, last_name, job_id, hire_date AS start_date
FROM employees;
</pre>

![task-1.6.png](https://drive.google.com/uc?id=1m-kSw4rSwQYcgZ2tzo5tXEEwaxgjhtSa)

In [ ]:
%%sql
# Insert Your Code Here

**Task 1.7** The HR department wants a query to display all unique job IDs from the <code>employees</code> table.

<pre>
SELECT DISTINCT job_id
FROM employees;
</pre>

![task-1.7.png](https://drive.google.com/uc?id=1dVJOQBA7Nv2f6ppF4driH-IkO15ix2ET)

In [ ]:
%%sql
# Insert Your Code Here

##**Lab 2**: Restricting and Sorting Data
This practice covers the following topics:
- Selecting data and changing the order of the rows that are displayed
- Restricting rows by using the <code>WHERE</code> clause
- Sorting rows by using the <code>ORDER BY</code> clause

In this practice, you build more reports by using statements that use the <code>WHERE</code> clause and the <code>ORDER BY</code> clause.

**Task 2.1** The HR department needs your assistance in creating some queries. Because of budget issues, the HR department needs a report that displays the last name and salary of employees who earn more than $12,000.

![task-2.1.png](https://drive.google.com/uc?id=1Jf76ZzpBJdmhq-HAGMbpv9WUc3ha3o0P)

In [ ]:
%%sql
# Insert Your Code Here

**Task 2.2** Create a report that displays the last name and department
number for employee number 176.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1AQWbDWAv5JxwgMZUdrn8wnrCV128J3Vd)


In [ ]:
%%sql
# Insert Your Code Here

**Task 2.3** The HR department needs to find high-salary and low-salary employees. Modify your SQL code created for Task 2.1 to display the last name and salary for any employee whose salary is not in the range \$5,000 through \$12,000.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1db8kCabhxb-spQdWJUToWb-jFMl9cjFF)


In [ ]:
%%sql
# Insert Your Code Here

**Task 2.4** Create a report to display the last name and hire date for employees with the last names of Matos and Talyor. Order the query in ascending order by hire date.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1i-XUEew_xIG20grV9i79rNlRtrIN_feS)

In [ ]:
%%sql
# Insert Your Code Here

**Task 2.5** Display the last name and department number of all employees in departments <code>20</code> or <code>50</code> in ascending order by employee number.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1bPzYwvT4xiGo6GA8XzA1HZ3z63So8xol)


In [ ]:
%%sql
# Insert Your Code Here

**Task 2.6** Modify the query for Task 2.3 to display the employee number and salary of employees who earn between \$5,000 and \$12,000. Label the columns <code>employee</code> and <code>monthly salary</code>, respectively.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1Cf5QtPmDv-KEXvarRAxHmbOdwL4fBKL6)


In [ ]:
%%sql
# Insert Your Code Here

**Task 2.7** The HR department needs a report that displays the last name and hire date of all employees who were hired in 2000.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1v2ATOUdIwMYbgCKxUn1g9NtIpLX-WUud)


In [ ]:
%%sql
# Insert Your Code Here

**Task 2.8** Create a report to display the employee number and job title of all employees who are currently a manager.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1tQjtbifEkssAz36RFss-biY50SzE2lSY)


In [ ]:
%%sql
# Insert Your Code Here

**Task 2.9** Create a report to display the employee number and salaryof all current employees. Sort the data in descending order of salary. Use the column’s numeric position in the ORDER BY clause.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1kdZAF91gOnKJgg2F2JgwhGLuJm1kh-tE)


In [ ]:
%%sql
# Insert Your Code Here

##**Lab 3**: Using Single-Row Functions to Customize Output
This practice covers the following topics:
- Writing a query that displays the current date
- Creating queries that require the use of numeric, character, and date functions
- Performing calculations of years and months of service for an employee

This practice provides a variety of exercises using the different functions that are available for
character, number, and date data types. Remember that for nested functions, the results are evaluated from the innermost function to the outermost function.

**Task 3.1** Write a query to display the system date. Label the column Date.

Note: If your database is remotely located in a different time zone, the output will be the date for the operating system on which the database resides.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1rx-d5wM7AANnaGV1TEQyftIiGkn9I8sV)

In [ ]:
%%sql
# Insert Your Code Here

**Task 3.2** The HR department needs a report to display the employee ID, last name, salary, and salary increased by 15.5% (expressed as a whole number) for each employee. Label the column <code>new_salary</code>.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1zfeoFuQTDBaM457ZRv6VXGNEJOamuUaW)


In [ ]:
%%sql
# Insert Your Code Here

**Task 3.3** Modify your query in the Task 3.2 to add a column that subtracts the old salary from the new salary. Label the column <code>increase</code>.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1MkU_OL40W33l7dVQTjLBUy9SiWhsJ4Sp)


In [ ]:
%%sql
# Insert Your Code Here

**Task 3.4** Write a query that displays the last name (with the first letter in uppercase and all the
other letters in lowercase) and the length of the last name for all employees whose name starts with the letters “J,” “A,” or “M.” Give each column an appropriate label. Sort the results by the employees’ last names.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1bblPebmjozZKz92sxfmrgu8aQYnb__Ir)


In [ ]:
%%sql
# Insert Your Code Here

##**Lab 4**: Using Conversion Functions and Conditional Expressions
This practice covers the following topics:
- Creating queries that use <code>CAST</code> function.
- Creating queries that use conditional expressions such as <code>CASE</code> and <code>IFNULL</code>.

**Task 4.1** Create a report that produces the following for each employee:
<code><employee last name></code> earns <code>salary</code> monthly but wants <code>3 times salary</code>. Label the column <code>dream_salary</code>.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1w05zy2kB8TB2CHBH7VBnQ9rPIYhAJhYh)


In [ ]:
%%sql
# Insert Your Code Here

**Task 4.2** Create a query that displays employees’ last names and commission amounts. If an
employee does not earn commission, show “No Commission.” Label the column <code>comm</code>.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1sqnC1mMpRAHZIhIz5aibcZlVOa3lqsj1)

In [ ]:
%%sql
# Insert Your Code Here

**Task 4.3** . Rewrite the statement in the preceding exercise by using the <code>CASE</code> syntax.

![hr-rl-sm.gif](https://drive.google.com/uc?id=16DnEozDjo58rwggN8Ug23yLkG6Iimgep)

In [ ]:
%%sql
# Insert Your Code Here

##**Lab 5**: Reporting Aggregated Data Using the Group Functions
This practice covers the following topics:
- Writing queries that use the group functions
- Grouping by rows to achieve multiple results
- Restricting groups by using the <code>HAVING</code> clause

After completing this practice, you should be familiar with using group functions and selecting
groups of data.

**Task 5.1** Determine the validity of the following statements. Either True or False.
- Group functions work across many rows to produce one result per group. <code>True/False</code>
- Group functions include nulls in calculations. <code>True/False</code>
- The <code>WHERE</code> clause restricts rows before inclusion in a group calculation.<code>True/False</code>

In [ ]:
%%sql
# Insert Your Code Here

**Task 5.2** The HR department needs the following reports (5.2-5.5): Find the highest, lowest, sum, and average salary of all employees. Label the columns <code>maximum</code>, <code>minimum</code>, <code>sum</code>, and <code>average</code>, respectively. Round your results to the
nearest whole number.


![hr-rl-sm.gif](https://drive.google.com/uc?id=1aiUYt_7_mXxppPZSporFGWgYtcyT6YKG)


In [ ]:
%%sql
# Insert Your Code Here

**Task 5.3** Modify the query in Task 5.2 to display the minimum, maximum, sum and average salary for each job type.


![hr-rl-sm.gif](https://drive.google.com/uc?id=1RLqrmRzwwIx3YkmEP6_UqACUAGt-BGlF)


In [ ]:
%%sql
# Insert Your Code Here

**Task 5.4** Write a query to display the number of people with the same job.


![hr-rl-sm.gif](https://drive.google.com/uc?id=1meBgSUvu7GhkPiS03cw2AKtXVEnT8sZQ)


In [ ]:
%%sql
# Insert Your Code Here

**Task 5.5** Find the difference between the highest and lowest salaries. Label the column <code>difference</code>.


![hr-rl-sm.gif](https://drive.google.com/uc?id=1763WAWmYkX_biNAFVsZnnElS7mrzLwRF)


In [ ]:
%%sql
# Insert Your Code Here

##**Lab 6**: Displaying Data from Multiple Tables Using Joins
This practice covers the following topics:
- Joining tables using an equijoin
- Performing outer and self-joins
- Adding conditions

This practice is intended to give you experience in extracting data from multiple tables using the
SQL:1999–compliant joins.

**Task 6.1** Write a query for the HR department to produce the addresses of all the departments. Use
the <code>locations</code> and <code>countries</code> tables. Show the location ID, street address, city, state or province, and country in the output. Use a <code>INNER JOIN</code> to produce the results.


![hr-rl-sm.gif](https://drive.google.com/uc?id=1lkDiZyPLys5zOeJ0Sacys-KGBTYQzkXB)


In [ ]:
%%sql
# Insert Your Code Here

**Task 6.2** The HR department needs a report of all employees with corresponding departments. Write a query to display the last name, department number, and department name for these employees.


![hr-rl-sm.gif](https://drive.google.com/uc?id=1JTsOIByK3ZfNh12_1lgvXZvxxrW-ku_Q)


In [ ]:
%%sql
# Insert Your Code Here

**Task 6.3** The HR department needs a report of employees in Toronto. Display the last name, job, department number, and the department name for all employees who work in Toronto.


![hr-rl-sm.gif](https://drive.google.com/uc?id=1-jT63i8B447qKEcAaTc_toW2LRVxd3ai)


In [ ]:
%%sql
# Insert Your Code Here

**Task 6.4** Create a report to display employees’ last names and employee numbers along with their
managers’ last names and manager numbers. Label the columns <code>employee</code>, <code>emp#</code>, <code>manager</code>, and <code>mgr#</code>, respectively.


![hr-rl-sm.gif](https://drive.google.com/uc?id=1VeZUVswgKiMid3cRt1EVGfo56lyOL2iQ)


In [ ]:
%%sql
# Insert Your Code Here

**Task 6.5** Modify the query used in Task 6.4 to display all employees, including King, who has no manager. Order the results by employee number.


![hr-rl-sm.gif](https://drive.google.com/uc?id=1sQPiQkfXGbxDYxz5gtfV7maZm84fXzeQ)


In [ ]:
%%sql
# Insert Your Code Here

**Task 6.6** Create a report for the HR department that displays employee last names, department
numbers, and all the employees who work in the same department as a given employee. Give each column an appropriate label.


![hr-rl-sm.gif](https://drive.google.com/uc?id=1hv0PjVrB9kFGAeNIehCbENfPdv9g3My4)


In [ ]:
%%sql
# Insert Your Code Here

**Task 6.7** The HR department needs a report on job grades and salaries. To familiarize yourself with
the <code>jobs</code> table, first show the structure of the <code>jobs</code> table. Then create a query that displays the name, job, department name, salary, mininum salary, and maximum salary for the job title for all employees.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1gZY5CRjpTTr95GadNgJluVi-rOs53w7i)

![hr-rl-sm.gif](https://drive.google.com/uc?id=18hzavaby0gGuzvB_blJkUDS199s4vUa3)

In [ ]:
%%sql
# Insert Your Code Here

In [ ]:
%%sql
# Insert Your Code Here

##**Lab 7**: Using Subqueries to Solve Queries
This practice covers the following topics:
- Using subqueries to find values that exist in one set of data and not in another

In this practice, you write complex queries using nested <code>SELECT</code> statements. For practice questions, you may want to create the inner query first. Make sure that it runs and produces the data that you anticipate before you code the outer query.

**Task 7.1** Create a report that displays the employee number, last name, and salary of all employees who earn more than the average salary. Sort the results in ascending order by salary.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1lDqcmDmbpJqFw5X6K3_iyZ6NLa4HUtPR)


In [ ]:
%%sql
# Insert Your Code Here

**Task 7.2** Write a query that displays the employee number and last name of all employees who work in a department with any employee whose last name contains the letter “u.”

![hr-rl-sm.gif](https://drive.google.com/uc?id=1lyGF5Ke37SsVMPPcThBmvCBAAyGCYLaP)


In [ ]:
%%sql
# Insert Your Code Here

**Task 7.3** The HR department needs a report that displays the last name, department number, and job ID of all employees whose department location ID is 1700.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1xFaRQxhRkPmZvd9LTblRFmaWxpZn3J83)


In [ ]:
%%sql
# Insert Your Code Here

**Task 7.4** Create a report for HR that displays the last name and salary of every employee who reports to King.

![hr-rl-sm.gif](https://drive.google.com/uc?id=19BGt-2C6DvPY-i0QynNRCUrwX8TE8Dj1)


In [ ]:
%%sql
# Insert Your Code Here

**Task 7.5** Create a report for HR that displays the department number, last name, and job ID for every employee in the Executive department.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1suXfsXKLaP7TVj6UmsQt6mSrT2Ng4YCM)


In [ ]:
%%sql
# Insert Your Code Here

**Task 7.6** Create a report that displays a list of all employees whose salary is more than the salary of any employee from department 60.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1fhQY6pcXvWRofbaWmg6QZo7Aqb2TXAWi)


In [ ]:
%%sql
# Insert Your Code Here

##**Lab 8**: Using the Set Operators
In this practice, you create reports by using the following:
- <code>UNION</code> operator
- <code>INTERSECT</code> operator
- <code>EXCEPT</code> operator

In this practice, you write queries using the set operators <code>UNION</code>, <code>INTERSECT</code>, and <code>EXCEPT</code>.

**Task 8.1** The HR department needs a list of department IDs for departments that do not contain the job ID <code>ST_CLERK</code>. Use the set operators to create this report.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1LO_v8kgF0S8UbeNeA5eeUp9LUohseCbJ)


In [ ]:
%%sql
# Insert Your Code Here

**Task 8.2** The HR department needs a list of countries that have no departments located in them. Display the country IDs and the names of the countries. Use the set operators to create this report.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1LRHEJ8xyL1l2kZRK52EjKJwZMquAc2ko)


In [ ]:
%%sql
# Insert Your Code Here

**Task 8.3** Produce a list of jobs for departments 10, 50, and 20, in that order. Display the job ID and department ID by using the set operators.


![hr-rl-sm.gif](https://drive.google.com/uc?id=1boH6TdPEif3k3mzWbeiqCTQDxmd3MNN8)


In [ ]:
%%sql
# Insert Your Code Here

**Task 8.4** Create a report that lists the employee IDs and job IDs of those employees who currently have a job title that is the same as their previous one (that is, they changed jobs but have now gone back to doing the same job they did previously).


![hr-rl-sm.gif](https://drive.google.com/uc?id=1Zyoa2HI8TgeMTWecxsl-3vkhqYD29YVy)


In [ ]:
%%sql
# Insert Your Code Here

**Task 8.5** The HR department needs a report with the following specifications:
- Last names and department IDs of all employees from the <code>employees</code> table,
regardless of whether or not they belong to a department
- Department IDs and department names of all departments from the <code>departments</code>
table, regardless of whether or not they have employees working in them.


![hr-rl-sm.gif](https://drive.google.com/uc?id=1QyBB-otpXkm11XEFdXCJWZB4G1MwUH6L)


In [ ]:
%%sql
# Insert Your Code Here

##**Lab 9**: Managing Tables Using DML Statements
This practice covers the following topics:
- Inserting rows into tables
- Updating and deleting rows in a table
- Controlling transactions

The HR department wants you to create SQL statements to insert, update, and delete employee data.

**Task 9.1** Create a table callled <code>my_employee</code> table following the specifiction below.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1crmGS6rbj-VfKSndgiSXKAsHK3CbS2yI)


In [ ]:
%%sql
# Insert Your Code Here

**Task 9.2** Describe the structure of the <code>my_employee</code> table to identify the column names.

In [ ]:
%%sql
# Insert Your Code Here

**Task 9.3** Create an <code>INSERT</code> statement to add the first row of data to the <code>my_employee</code> table from the following sample data. Do not list the columns in the <code>INSERT</code> clause. *Do not enter all rows yet.*

![hr-rl-sm.gif](https://drive.google.com/uc?id=18P-spYgHY90XjsycdgzRvh3-GAy-ynzB)

In [ ]:
%%sql
# Insert Your Code Here

**Task 9.4** Populate the <code>MY_EMPLOYEE</code> table with the second row of the sample data from the preceding list. This time, list the columns explicitly in the INSERT clause.

In [ ]:
%%sql
# Insert Your Code Here

**Task 9.5** Confirm your addition to the table.

![hr-rl-sm.gif](https://drive.google.com/uc?id=18iNa3cFtj1-gH2cMAen5B3cOczh4NCoj)


In [ ]:
%%sql
# Insert Your Code Here

**Task 9.6** Write an <code>INSERT</code> statement to load the remaining rows into the <code>my_employee</code> table.

In [ ]:
%%sql
# Insert Your Code Here

**Task 9.7** Make the data additions permanent. Confirm  your addition to the table.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1zn3iztWf7DMl6YNA0vQDrZZm0yVs89XQ)


In [ ]:
%%sql
# Insert Your Code Here

In [ ]:
%%sql
# Insert Your Code Here

**Task 9.8** Update and delete data in the <code>my_employee</code> table. Change the last name of employee 4 to Johnson.

In [ ]:
%%sql
# Insert Your Code Here

**Task 9.9** Change the salary to \$1,000 for all employees who have a salary less than \$900.

In [ ]:
%%sql
# Insert Your Code Here

**Task 9.10** . Verify your changes to the table.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1QeiGz5E42bYS3Xjxosr9-cjAaXhGaelH)


In [ ]:
%%sql
# Insert Your Code Here

**Task 9.11** Delete Isaac Newton from the <code>my_employee</code> table.

In [ ]:
%%sql
# Insert Your Code Here

**Task 9.12** Confirm your changes to the table.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1ATgGpKeAIuc2Wi8Gytg2fkcPpktnt3r6)


In [ ]:
%%sql
# Insert Your Code Here

**Task 9.13** Make the change permanent.

In [ ]:
%%sql
# Insert Your Code Here

**Task 9.14** Control the data transaction to the my_employee table by using <code>TRANSACTION</code> block, <code>SAVEPOINT</code> and <code>ROLLBACK</code>.
<pre>
START TRANSACTION;
SAVEPOINT sp;
INSERT INTO my_employee VALUES (1, 'Newton', 'Isaac', 'inewton', 886);
ROLLBACK TO SAVEPOINT sp;
</pre>

In [ ]:
%%writefile ./transaction_block.sql
START TRANSACTION;
SAVEPOINT sp;
INSERT INTO my_employee VALUES (1, 'Newton', 'Isaac', 'inewton', 886);
ROLLBACK TO SAVEPOINT sp;

In [ ]:
# Insert Your Code Here

**Task 9.15** Confirm the change has been reversed.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1j8O3VpMU7acjx702vU3vrFQKJIaVRzKZ)


In [ ]:
%%sql
# Insert Your Code Here

**Task 9.13** Drop the <code>my_employee</code> table, and confirm the table was dropped. Confirm the table was dropped.

In [ ]:
%%sql
# Insert Your Code Here

In [ ]:
%%sql
# Insert Your Code Here

##**Lab 10**: Data Definition Language (DDL)
This practice covers the following topics:
- Creating new tables
- Creating a new table by using the <code>CREATE TABLE AS</code> syntax
- Verifying that tables exist
- Altering tables
- Adding columns
- Dropping columns
- Setting a table to read-only status
- Dropping tables

Create new tables by using the <code>CREATE TABLE</code> statement. Confirm that the new table was
added to the database. You also learn to set the status of a table as <code>READ ONLY</code>, and then
revert to <code>READ/WRITE</code>. You use the <code>ALTER TABLE</code> command to modify table columns.

**Task 10.1** Create the <code>dept</code> table based on the following table instance chart. Execute the statement in the script to create the
table. Confirm that the table is created.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1osJcrerBWa4KVGC0C48w6bhDdPBk7Yqe)


In [ ]:
%%sql
# Insert Your Code Here

In [ ]:
%%sql
# Insert Your Code Here
DESCRIBE dept;

**Task 10.2** Create the <code>emp</code> table based on the following table instance chart. Execute the statement in the script to create the table. Confirm that the table is created.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1P4anm5udxMkRBYjl6-fUTYqtblf5U-pt)


In [ ]:
%%sql
# Insert Your Code Here

In [ ]:
%%sql
# Insert Your Code Here

**Task 10.3** Modify the <code>emp</code> table. Add a <code>commission</code> column of the <code>DECIMAL</code> data type, with precision 2
and scale 2. Confirm your modification.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1Tzw1qXsNDDAKIkBYUdVY1dWNHXBzX_i9)


In [ ]:
%%sql
# Insert Your Code Here

In [ ]:
%%sql
# Insert Your Code Here

**Task 10.4** Modify the emp table to allow for longer employee last names. Confirm your modification.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1K2sKSgfuG5vGbk0D6h6UEz6C0py-vE38)


In [ ]:
%%sql
# Insert Your Code Here

In [ ]:
%%sql
# Insert Your Code Here

**Task 10.5** Drop the <code>first_name</code> column from the <code>emp</code> table. Confirm your modification by checking the description of the table.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1RfZyDgKKoer6Otm0ZsmARPwGnRephF5r)


In [ ]:
%%sql
# Insert Your Code Here

In [ ]:
%%sql
# Insert Your Code Here

**Task 10.6** Create the </code>employee2</code> table based on the structure of the <code>employees</code> table. Include only the <code>employee_id</code>, <code>first_name</code>, <code>last_name</code>, <code>salary</code>, and <code>department_id</code> columns. Name the columns in your new table <code>id</code>, <code>first_name</code>, <code>last_name</code>, <code>SALARY</code>, and <code>dept_id</code>, respectively.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1MK57dhqzWHv-WjnD7t0UJFIEckDzZdh4)


In [ ]:
%%sql
# Insert Your Code Here

In [ ]:
%%sql
# Insert Your Code Here

**Task 10.7** Drop the <code>my_employee</code>, <code>emp</code>, <code>dept</code>, and <code>employee2</code> table. Confirm the tables were dropped.

In [ ]:
%%sql
# Insert Your Code Here

In [ ]:
%%sql
# Insert Your Code Here

##**Lab 11**: Creating Views
This practice covers the following topics:
- Creating a simple view
- Creating a complex view
- Creating a view with a check constraint
- Attempting to modify data in the view
- Querying the dictionary views for view information
- Removing views


**Task 11.1** The staff in the HR department wants to hide some of the data in the <code>employees</code> table. Create a view called <code>employees_vw</code> based on the employee numbers, employee last names, and department numbers from the <code>employees</code> table. The heading for the
employee name should be <code>employee</code>.


In [ ]:
%%sql
# Insert Your Code Here

**Task 11.2** Confirm that the view works. Display the contents of the <code>employees_vw</code> view.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1ikRx9DlY7hExvRp4a3SvsL7KZXFKrBk2)

In [ ]:
%%sql
# Insert Your Code Here

**Task 11.3** Using your <code>employees_vw</code> view, write a query for the HR department to display all employee names and department numbers.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1NIyw80Sguiui0E5SDcP73KGrfCF0hAaM)

In [ ]:
%%sql
# Insert Your Code Here

**Task 11.4** Department 80 needs access to its employee data. Create a view named <code>dept80</code> that contains the employee numbers, employee last names, and department numbers for all employees in department 80. They have requested that you label the view columns <code>empno</code>,
<code>employee</code> and <code>depno</code>. For security purposes, do not allow an employee to be reassigned to another department through the view.


In [ ]:
%%sql
# Insert Your Code Here

**Task 11.5** Display the structure and contents of the <code>dept80</code> view.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1wPlmKDWzRwVV8JXlCZyXdswq3KYO1bqA)

![hr-rl-sm.gif](https://drive.google.com/uc?id=1PRtFcge2Jn19pJDeZ1vhZXwefTmoDTXG)


In [ ]:
%%sql
# Insert Your Code Here

In [ ]:
%%sql
# Insert Your Code Here

**Task 11.6** Test your view. Attempt to reassign <code>Abel</code> to department 50. The error is because the <code>dept50</code> view has been created with the WITH <code>CHECK OPTION</code> constraint. This ensures that the <code>depno</code> column in the view is protected from being changed.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1wBXmJceA6oZlJsuHbHZ9deAfTNX4fnJH)


In [ ]:
%%sql
# Insert Your Code Here

**Task 11.7** Remove the views created in this practice. Confirm if the views have been removed.

In [ ]:
%%sql
# Insert Your Code Here

In [ ]:
%%sql
# Insert Your Code Here

##**Lab 12**: Retrieving Data by Using Subqueries
This practice covers the following topics:
- Creating multiple-column subqueries
- Writing correlated subqueries
- Using the <code>EXISTS</code> operator
- Using scalar subqueries
- Using the <code>WITH</code> clause

**Task 12.1** Write a query to display the last name, department number, and salary of any employee whose department number and salary both match the department number and salary of any employee who earns a commission.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1W9MCup-0A3SrmOtMbmLk6lQGtQGDzO1S)

In [ ]:
%%sql
# Insert Your Code Here

**Task 12.2** Display the last name, department name, and salary of any employee whose salary and <code>job_id</code> match the salary and job_ID of any employee located in location ID 1700.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1KaGyD5dNRgedEOhh465XGxmxs2GBqLqd)


In [ ]:
%%sql
# Insert Your Code Here

**Task 12.3** Create a query to display the last name, hire date, and salary for all employees who have the same salary and <code>manager_id</code> as <code>Kochhar</code>.

Note: Do not display <code>Kochhar</code> in the result set.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1Vw_PbSRY7UX_QMeTGyYFPvZShSDEudM3)



In [ ]:
%%sql
# Insert Your Code Here

**Task 12.4** Create a query to display the employees who earn a salary that is higher than the salary of all the sales managers (<code>JOB_ID = 'SA_MAN'</code>). Sort the results on salary from the highest to the lowest.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1PU560ftaHz6EsDtcS3n-v7mor8kXpQb6)


In [ ]:
%%sql
# Insert Your Code Here

**Task 12.5** Display details such as the employee ID, last name, and department ID of those employees who live in cities the names of which begin with <code>T</code>.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1p3ubDNkUeeYf614BChU2319Fz1gz1O5V)


In [ ]:
%%sql
# Insert Your Code Here

**Task 12.6** Write a query to find all employees who earn more than the average salary in their departments. Display last name, salary, department ID, and the average salary for the department. Sort by average salary and round to two decimals. Use aliases for the columns retrieved by the query as shown in the sample output.

![hr-rl-sm.gif](https://drive.google.com/uc?id=10G2EzeADo2oE88N2xvBGwyInGuTuZYEH)


In [ ]:
%%sql
# Insert Your Code Here

**Task 12.7** Find all employees who are not supervisors.
- First, do this by using the NOT EXISTS operator.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1xkZ2jD9nKO8Ag3xXTh_9InPvNZvo9XlS)

In [ ]:
%%sql
# Insert Your Code Here

- Can this be done by using the NOT IN operator? How, or why not?

![hr-rl-sm.gif](https://drive.google.com/uc?id=1Ogn4fc3e-QyVnusf22KQLn2d-8_uL5-e)


In [ ]:
%%sql
# Insert Your Code Here

**Task 12.8** Write a query to display the last names of the employees who earn less than the average salary in their departments.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1a3fzbx7ENlNiXxzKvoo5zHp7y-smC7tv)


In [ ]:
%%sql
# Insert Your Code Here

**Task 12.9** Write a query to display the last names of employees who have one or more coworkers in their departments with later hire dates but higher salaries.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1TU42V2wmlkyDAKvKb8DHqouOvUIOAcl3)


In [ ]:
%%sql
# Insert Your Code Here

**Task 12.10** Write a query to display the employee ID, last names, and department names of all employees.

![hr-rl-sm.gif](https://drive.google.com/uc?id=1E5AvzSRVMKXArDN1Zj7ayREoz2-GU6W5)


In [ ]:
%%sql
# Insert Your Code Here

**Task 12.11** Write a query to display the department names of those departments whose total salary cost is above one-eighth (1/8) of the total salary cost of the whole company. Use the <code>WITH</code> clause to write this query.

![hr-rl-sm.gif](https://drive.google.com/uc?id=12xS3asVbXZUaF-VPgsF8nhMoXd6A0L3r)


In [ ]:
%%sql
# Insert Your Code Here